Lib/test/test_ast/test_ast.py

```diff
@@ -3035,3 +3035,135 @@ def test_cli_file_input(self):
         self.assertEqual(expected.splitlines(),
                          res.out.decode("utf8").splitlines())
         self.assertEqual(res.rc, 0)
+
+
+class ASTOptimiziationTests(unittest.TestCase):
+    binop = {
+        "+": ast.Add(),
+        "-": ast.Sub(),
+        "*": ast.Mult(),
+        "/": ast.Div(),
+        "%": ast.Mod(),
+        "<<": ast.LShift(),
+        ">>": ast.RShift(),
+        "|": ast.BitOr(),
+        "^": ast.BitXor(),
+        "&": ast.BitAnd(),
+        "//": ast.FloorDiv(),
+        "**": ast.Pow(),
+    }
+
+    unaryop = {
+        "~": ast.Invert(),
+        "+": ast.UAdd(),
+        "-": ast.USub(),
+    }
+
+    def find_node(self, code, optimize, target):
+        tree = ast.parse(code, optimize=optimize)
+        for node in ast.walk(tree):
+            if ast.compare(node, target):
+                return True
+        return False
+
+    def create_binop(self, operand):
+        return ast.BinOp(left=ast.Constant(1), op=self.binop[operand], right=ast.Constant(1))
+
+    def create_unaryop(self, operand):
+        return ast.UnaryOp(op=self.unaryop[operand], operand=ast.Constant(1))
+
+    def assert_ast(self, code, non_optimized_target, optimized_target):
+        # Searching for a non-optimized node in an unoptimized AST
+        self.assertTrue(self.find_node(code, optimize=-1, target=non_optimized_target))
```

Comment:
Would it be possible to modify find_node() to fail with a message message like "cannot find xxx in xxx"? So the AST is displayed on error to ease debugging. (Need to support the opposite for assertFalse() below.)

Reply:
Thanks, that would be really useful. Here how it would look:
```python
AssertionError: False is not true : Cannot find Subscript(value=Tuple(elts=[Constant(value=1)], ctx=Load()), slice=Constant(value=1), ctx=Load()) in Module(body=[Expr(value=Subscript(value=Tuple(elts=[Constant(value=1)], ctx=Load()), slice=Constant(value=0), ctx=Load()))])
```